include/tscpp/util/Comparable.h

```diff
@@ -0,0 +1,196 @@
+  auto
+  ComparableFunction(T const &lhs, U const &rhs, meta::CaseTag<2> const &) -> decltype(cmp(rhs, lhs), int())
+  {
+    return static_cast<int>(cmp(rhs, lhs)) * -1;
+  }
+
+  /// Use the @c T::cmp method.
+  template <typename T, typename U>
+  auto
+  ComparableFunction(T const &lhs, U const &rhs, meta::CaseTag<1> const &) -> decltype(lhs.cmp(rhs), int())
+  {
+    return static_cast<int>(lhs.cmp(rhs));
+  }
+
+  /// Use @c U::cmp and flip the result.
+  template <typename T, typename U>
+  auto
+  ComparableFunction(T const &lhs, U const &rhs, meta::CaseTag<0> const &) -> decltype(rhs.cmp(lhs), int())
+  {
+    return static_cast<int>(rhs.cmp(lhs)) * -1;
+  }
+
+} // namespace detail
+
+/** Create standard comparison operators given a compare function.
+ *
+ * The operators supported are @c == @c != @c \< @c \> @c \<= @c \>=
+ *
+ * To successfully use this mixin, there are two requirements.
+ * - There must be a ternary comparison that returns an int in the standard ternary compare style.
+ * - The class must inherit from this mixin.
+ *
+ * The standard ternary compare must return an @c int which is
+ * - negative if @a lhs is smaller than @a rhs
+ * - 0 if @a lhs is equal to @a rhs
+ * - positive if @a lhs is greater than @a rhs
+ *
+ * If a comparison operator is used and at least one of the operands inherits from @c Comparable
+ * then the types are probed for ternary comparisons. The order is
+ *
+ * - @c ts::ComparablePolicy<lhs,rhs> specialization
+ * - function @c cmp(lhs,rhs)
+ * - function @c cmp(rhs,lhs)
+ * - method @c lhs::cmp(rhs)
+ * - method @c rhs::cmp(lhs)
+ *
+ * Once a class inherits from this mixin, then comparisons are supported against any type for
+ * which a ternary compare can be found.
+ *
+ * @code
+ * class T : public ts::Comparable
+ * @endcode
+ *
+ * To provide self comparison operators, it would suffice to have
+ *
+ * @code
+ *   int cmp(T const& that) const;
+ * @endcode
+ *
+ * If comparison operators against @c std::string_view should supported, this could be done by adding
+ *
+ * @code
+ *   int cmp(std::string_view that) { return strcmp(text, that); }
+ * @endcode
+ *
+ * If both of these are present, then the following are now valid
+ * @code
+ * T t, t1, t2;
+ * if ("walt"sv == t) {...}
+ * if ("walt" == t) {...} // because string literals convert to string_view
+ * if (t1 < t2) {...}
+ * @endcode
+ */
+struct Comparable {
+};
+
+// ---
+// For each comparison operator, a template overload is provided. These overloads are enabled for
+// overload resolution iff at least one operand inherits from @c ts::Comparable. In that case the
+// @c ComparableFunction machinery is engaged to probe for a ternary comparison. All of this should
+// optimize away after compilation.
+
+template <typename T, typename U>
+auto
+operator==(T const &lhs, U const &rhs) ->
+  typename std::enable_if<std::is_base_of<Comparable, T>::value || std::is_base_of<Comparable, U>::value, bool>::type
```

Comment:
Consider using std::is_convertable in place of std::is_base_of .  This allows the alternative of tagging a class for comparison using a dummy conversion operator:

operator Comparable ();

This avoids the ambiguous base class error if both a base class and a class derived from it need to be tagged as comparable.  It also allows a derived class to suppress inheritance of the comparison operators of a base class with:

operator Comparable () = delete;

https://godbolt.org/z/ToweoI